extensions/ql-vscode/src/queries-panel/query-discovery.ts

```diff
@@ -1,147 +1,106 @@
 import { QueryDiscoverer } from "./query-tree-data-provider";
-import { extLogger } from "../common";
+import { QueryPackDiscovery } from "./query-pack-discovery";
+import { isEqualOrHigherPath } from "../pure/helpers-pure";
+import { FilePathDiscovery } from "./file-path-discovery";
 
-/**
- * The results of discovering queries.
- */
-export interface QueryDiscoveryResults {
-  /**
-   * A tree of directories and query files.
-   * May have multiple roots because of multiple workspaces.
-   */
-  queries: Array<FileTreeDirectory<string>>;
-
-  /**
-   * File system paths to watch. If any ql file changes in these directories
-   * or any subdirectories, then this could signify a change in queries.
-   */
-  watchPaths: Uri[];
+interface Query {
+  path: string;
+  language: string | undefined;
 }
 
 /**
  * Discovers all query files contained in the QL packs in a given workspace folder.
  */
 export class QueryDiscovery
-  extends Discovery<QueryDiscoveryResults>
+  extends FilePathDiscovery<Query>
   implements QueryDiscoverer
 {
-  private results: QueryDiscoveryResults | undefined;
+  private readonly queryPackDiscovery: QueryPackDiscovery;
 
-  private readonly onDidChangeQueriesEmitter: AppEventEmitter<void>;
-  private readonly watcher: MultiFileSystemWatcher = this.push(
-    new MultiFileSystemWatcher(),
-  );
+  constructor(private readonly app: App, cliServer: CodeQLCliServer) {
+    super("Query Discovery", "**/*.ql", app);
 
-  constructor(
-    private readonly app: App,
-    private readonly cliServer: CodeQLCliServer,
-  ) {
-    super("Query Discovery", extLogger);
+    this.queryPackDiscovery = this.push(new QueryPackDiscovery(app, cliServer));
+    this.push(
+      this.queryPackDiscovery.onDidChangeQueryPacks(
+        this.recomputeAllQueryLanguages.bind(this),
+      ),
+    );
+  }
 
-    this.onDidChangeQueriesEmitter = this.push(app.createEventEmitter<void>());
-    this.push(app.onDidChangeWorkspaceFolders(this.refresh.bind(this)));
-    this.push(this.watcher.onDidChange(this.refresh.bind(this)));
+  public async initialRefresh() {
+    await this.queryPackDiscovery.initialRefresh();
+    return super.initialRefresh();
   }
 
-  public get queries(): Array<FileTreeDirectory<string>> | undefined {
-    return this.results?.queries;
+  public buildQueryTree(): Array<FileTreeNode<string>> {
+    const roots = [];
+    for (const workspaceFolder of getOnDiskWorkspaceFoldersObjects()) {
+      const queriesInRoot = this.paths.filter((query) =>
+        isEqualOrHigherPath(workspaceFolder.uri.fsPath, query.path),
+      );
+      if (queriesInRoot.length > 0) {
+        const root = new FileTreeDirectory<string>(
+          workspaceFolder.uri.fsPath,
+          workspaceFolder.name,
+          this.app.environment,
+        );
+        for (const query of queriesInRoot) {
+          const dirName = dirname(normalize(relative(root.path, query.path)));
+          const parentDirectory = root.createDirectory(dirName);
```

Comment:
It looks like this creates a hierarchy that is two deep: Workspace root folder -> query. Is there any indication of which query packs each query belongs to?

Will packs with no queries still be shown? Is this what we want?

Reply:
Yes, the top level entry is the workspace root folder. We don't directly indicate which pack a query comes from, but we display the language of the query next to it and we get that info from the query pack.
<img width="494" alt="Screenshot 2023-06-12 at 16 06 57" src="https://github.com/github/vscode-codeql/assets/3749000/2bc0b875-2c51-4514-a76e-3031eeef0b28">

> Will packs with no queries still be shown? Is this what we want?

I checked about this and it's not part of the plan to show query packs in the queries panel. So we won't directly show or link to them, and so we also won't show any packs that don't have queries in them. This is obviously something that could be changed in the future if it would be useful for users to have quick access to their query packs, without going back to the regular file explorer tab.